src/compute-client/src/controller/instance.rs

```diff
@@ -420,6 +420,11 @@ where
             .remove(&id)
             .expect("replica not found");
 
+        // In case the replica crashes and we receive a drop replica request
+        // at the same time, the cleanup request will race with the rehydration.
+        // Hence we also have to stop a pending rehydration request.
+        self.compute.failed_replicas.remove(&id);
```

Comment:
I think because of this possibility we should definitely move the orchestrator call to drop the service from the replica task into `remove_replica`. It might be that the replica task is not running right now because the replica is failed, in which case there would be nothing to drop the orchestrator service and we would leak it. That cannot happen if we `tokio::spawn` a cleanup task in `remove_replica`.

Reply:
Hm, then it would be better to also move the ensure service there, otherwise a quick `CREATE` followed by `DROP` could swap orders and the service is dropped before it's ensured.

Reply:
Good catch. I think this can happen in any case as long as ensure service and drop service are not executed in-line. Like, if we spawned a task in `add_replica` there would still be no guarantee that that gets scheduled before the task spawned by `remove_replica`.

Okay, alternative proposal: Run a rehydration of failed replicas at the beginning of `remove_replicas`, to ensure the task is actually there to stop the service in the orchestrator. Then I think we also don't need to worry about the race here.

Reply:
Hm running a rehydration only to drop it sounds not very elegant, especially if you want to drop a wedged replica, the task might always give up before you can reach it. If we want to do that, we should have the task wait until it receives a terminate command from Instance. We'd probably need an extra message type: TerminateWithOrchestratorRemove and TerminateWithoutOrchestratorRemoval...

At this point, I think it would be better to have an extra OrchestratorTask and an OrchestratorClient and they send each other messages. The task would serialize all the requests (ok then the initial refactor was a bit pointless, as the orchestrator calls could have stayed in the sequencer...)

Reply:
Hm, true, I didn't consider that the replica task always waits for a successful connection first.

Sounds like an orchestrator task really is the best solution then.

> (ok then the initial refactor was a bit pointless, as the orchestrator calls could have stayed in the sequencer...)
Wait, why would an orchestrator task mean we need to block the sequencer?